src/translate_rapidpro_keys.py

```diff
@@ -0,0 +1,111 @@
+                old_key = remapping.rapid_pro_key
+                new_key = remapping.pipeline_key
+                
+                if old_key in td and new_key not in td:
+                    remapped[new_key] = td[old_key]
+
+            td.append_data(remapped, Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+
+    @classmethod
+    def set_rqa_raw_keys_from_show_ids(cls, user, data, raw_id_map):
+        """
+        Despite the earlier phases of this pipeline stage using a common 'rqa_message' field and then a 'show_id'
+        field to identify which radio show a message belonged to, the rest of the pipeline still uses the presence
+        of a raw field for each show to determine which show a message belongs to. This function translates from
+        the new 'show_id' method back to the old 'raw field presence` method.
+        
+        TODO: Update the rest of the pipeline to use show_ids, and/or perform remapping before combining the datasets.
+
+        :param user: Identifier of the user running this program, for TracedData Metadata.
+        :type user: str
+        :param data: TracedData objects to set raw radio show message fields for.
+        :type data: iterable of TracedData
+        :param raw_id_map: Dictionary of show id to the rqa message key to assign each td[rqa_message} to.
+        :type raw_id_map: dict of int -> str
+        """
+        for td in data:
+            for show_id, message_key in raw_id_map.items():
+                if "rqa_message" in td and td.get("show_id") == show_id:
+                    td.append_data({message_key: td["rqa_message"]},
+                                   Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+
+    @classmethod
+    def translate_rapid_pro_keys(cls, user, data, pipeline_configuration, coda_input_dir):
+        """
+        Remaps the keys of rqa messages in the wrong flow into the correct one, and remaps all Rapid Pro keys to
+        more usable keys that can be used by the rest of the pipeline.
+        
+        TODO: Break this function such that the show remapping phase happens in one class, and the Rapid Pro remapping
+              in another?
+        """
+        # Set a show id field for each message, using the presence of Rapid Pro value keys in the TracedData.
+        # Show ids are necessary in order to be able to remap radio shows and key names separately (because data
+        # can't be 'deleted' from TracedData).
+        cls.set_show_ids(user, data, cls.SHOW_ID_MAP)
+
+        # Remap the keys used by Rapid Pro to more usable key names that will be used by the rest of the pipeline.
+        cls.remap_key_names(user, data, pipeline_configuration)
+
+        # Convert from the new show id format to the raw field format still used by the rest of the pipeline.
+        cls.set_rqa_raw_keys_from_show_ids(user, data, cls.RAW_ID_MAP)
```

Comment:
...and this if there is only one radio show?

Reply:
I was wondering... No I don't need it removed.